OAuth 2 artefact binding (#149)
oauth-wg/draft-ietf-oauth-attestation-based-client-auth: draft-ietf-oauth-attestation-based-client-auth.md

    target: "https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms"

informative:

>   RFC6749:

Author: ```suggestion
```

  RFC7523: RFC7523
  RFC9901: RFC9901
  ARF:
  	title: "The European Digital Identity Wallet Architecture and Reference Framework"

>   SD-JWT: I-D.ietf-oauth-selective-disclosure-jwt

Reviewer: I don't believe this is used anyhwere?
```suggestion
```

Authorization servers issuing a refresh token in response to a token request using the client attestation mechanism as defined by this draft MUST bind the refresh token to the Client Instance and its associated public key, and NOT just the client as specified in section 6 {{RFC6749}}. To prove this binding, the Client Instance MUST use the client attestation mechanism when refreshing an access token. The client MUST also use the same key that was present in the "cnf" claim of the client attestation that was used when the refresh token was issued.

## Binding of OAuth protocol artifacts

> Authorization servers using Attestation-Based Client Authentication are RECOMMENDED to bind relevant protocol artifacts to the Client Instance and its associated public key where possible, and NOT just the client as specified in section 6 {{RFC6749}}. Examples of these artifacts include but are not limited to:

Reviewer: ```suggestion
Authorization servers using Attestation-Based Client Authentication are RECOMMENDED to bind relevant protocol artifacts to the Client Instance and its associated public key where possible, and NOT just the client as specified in section 6 {{RFC6749}}. Note that this only applies if Attestation-Based Client Authentication is used as a Client Authentication. Examples of these artifacts include but are not limited to:
```

Reviewer 2: > and NOT just the client as specified in section 6 {{RFC6749}}

What are you referring to here? Section 6 is about refresh tokens.

Author: ```suggestion
Authorization servers using Attestation-Based Client Authentication are RECOMMENDED to bind relevant protocol artifacts to the Client Instance and its associated public key where possible, and NOT just the client as specified in {{RFC6749}}. Examples of these artifacts include but are not limited to:
```

Reviewer: ```suggestion
Authorization servers using Attestation-Based Client Authentication are RECOMMENDED to bind relevant protocol artifacts to the Client Instance and its associated public key where possible, and NOT just the client as specified in {{RFC6749}}. Note that this only applies if Attestation-Based Client Authentication is used as Client Authentication. Examples of these artifacts include but are not limited to:
```

Author: @aaronpk what we meant by this is that the protocol artifacts should not be bound to just the "client" as per ordinary RFC6749, as it would allow once instance of a client to leverage artifacts that were issued or assigned to another instance e.g a refresh token. I take you point though, so I've removed the explicit mention of section 6 here and generalised it to refer just to the client.

> - The authorization_code as specified in section 4.1 {{RFC6749}}.

Reviewer 2: ```suggestion
- The authorization_code as specified in {{Section 4.1 of RFC6749}}.
```

Authorization servers issuing a refresh token in response to a token request using the client attestation mechanism as defined by this draft MUST bind the refresh token to the Client Instance and its associated public key, and NOT just the client as specified in section 6 {{RFC6749}}. To prove this binding, the Client Instance MUST use the client attestation mechanism when refreshing an access token. The client MUST also use the same key that was present in the "cnf" claim of the client attestation that was used when the refresh token was issued.

## Binding of OAuth protocol artefacts

> Authorization servers where possible are RECOMMENDED to bind relevant protocol artefacts to the Client Instance and its associated public key, and NOT just the client as specified in section 6 {{RFC6749}}. Examples of these artefacts include but are not limited to:

Reviewer: we used artefact so far in the spec (only 1 place to be fair) - we should keep things constant

- The authorization_code as specified in section 4.1 {{RFC6749}}.
- The auth_req_id as specified in section 7.3 {{CIBA}}.

> How this binding is established and then proven is specific to the protocol artifact. For example establishing binding to an authorization_code involves the client instance using client attestation in the authorization request, and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint using an authorization code grant.

Author: ```suggestion
How this binding is established and then proven is specific to the protocol artefact. For example establishing binding to an authorization_code involves the client instance using client attestation in the authorization request, and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint using an authorization code grant.
```

> How this binding is established and then proven is specific to the protocol artefact. For example establishing binding to an authorization_code involves the client instance using client attestation in the authorization request, and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint using an authorization code grant.

Reviewer 3: ```suggestion
How this binding is established and then proven is specific to the protocol artifact. For example establishing binding to an authorization_code involves the client instance using client attestation in the authorization request, and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint using an authorization code grant.
```

Author: See above I think consensus is to use artefact.

> How this binding is established and then proven is specific to the protocol artifact. For example establishing binding to an authorization_code involves the client instance using client attestation in the authorization request, and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint using an authorization code grant.

Reviewer 4: It's not sensible to pass the attestation in the authorization request right? Although we have seen people try to pass private key jwt assertions to the authorization endpoint, so I think being a bit clearer here is good. Maybe something like this:

```suggestion
How this binding is established and then proven is specific to the protocol artifact. For example establishing binding to an authorization_code involves the client instance using client attestation before the user is redirected to the Authorization Endpoint (for example by using PAR, {{RFC9126}}), and proving binding of the authorization_code to the Client Instance involves using the client attestation mechanism to authenticate at the token endpoint when performing the authorization code grant.
```

Author: Yep agreed, good clarification.
